src/pybind/mgr/orchestrator.py

```diff
@@ -0,0 +1,266 @@
+
+
+class StatelessServiceSpec(object):
+    # Request to orchestrator for a group of stateless services
+    # such as MDS, RGW, nfs gateway, iscsi gateway
+    """
+    Details of stateless service creation.
+
+    This is *not* supposed to contain all the configuration
+    of the services: it's just supposed to be enough information to
+    execute the binaries.
+    """
+
+    def __init__(self):
+        self.placement = PlacementSpec()
+
+        # Give this set of statelss services a name: typically it would
+        # be the name of a CephFS pool, RGW zone, etc.  Must be unique
+        # within one ceph cluster.
+        self.name = ""
+
+        # Minimum and maximum number of service instances
+        self.min_size = 1
+        self.max_size = 1
+
+        # Arbitrary JSON-serializable object.
+        # Maybe you're using e.g. kubenetes and you want to pass through
+        # some replicaset special sauce for autoscaling?
+        self.extended = {}
+
+
+class InventoryFilter(object):
+    """
+    When fetching inventory, use this filter to avoid unnecessarily
+    scanning the whole estate.
+
+    Typical use: filter by host when presenting UI workflow for configuring
+                 a particular server.
+                 filter by label when not all of estate is Ceph servers,
+                 and we want to only learn about the Ceph servers.
+                 filter by label when we are interested particularly
+                 in e.g. OSD servers.
+
+    """
+    def __init__(self):
+        self.labels = None  # Optional: get info about nodes matching labels
+        self.hostnames = None  # Optional: get info about certain hosts only
+
+
+class InventoryBlockDevice(object):
+    """
+    When fetching inventory, block devices are reported in this format.
+
+    Note on device identifiers: the format of this is up to the orchestrator,
+    but the same identifier must also work when passed into StatefulServiceSpec
+
+    "Extended" is for reporting any special configuration that may have
+    already been done out of band on the block device.  For example, if
+    the device has already been configured for encryption, report that
+    here so that it can be indicated to the user.  The set of
+    extended properties may differ between orchestrators.  An orchestrator
+    is permitted to support no extended properties (only normal block
+    devices)
+    """
+    def __init__(self):
+        self.blank = False
+        self.type = None  # 'ssd', 'hdd', 'nvme'
+        self.id = None  # unique within a node (or globally if you like).
+        self.size = None  # byte integer.
+        self.extended = None  # arbitrary JSON-serializable object
+
+        # TODO: should we report a "size_avail" to indicate a partially
+        # used device (has some journal partitions but also room for more?)
+
+
+class InventoryNode(object):
+    def __init__(self):
+        self.hostname = None  # unique within cluster
```

Comment:
Some management tools might use an identifier different from the hostname to identify the node. Maybe we should have an `node_id` attribute and a `hostname` attribute.

Reply:
But all systems have hostnames, right?  We would only need the node_id in the case where hostnames weren't unique.

Are you thinking of things like the salt minion ID for use when users have left all their hostnames as localhost.localdomain?  I have kind of mixed feelings about how much we should try to support systems with non-unique hostnames, because they will end up looking crazy in the UI anyway when we show e.g. services grouped by host.

Reply:
Even if all machines have unique hostnames, which is the expected case, if we use the "hostname" then we need to maintain a map, inside the orchestrator module implementation, that maps the hostname and the id used by the orchestration system to identify that machine.

Also, how does it work in a containerized environment like k8s cluster? what's node in k8s cluster? is a pod, or physical machine? That's why I think that using the `node_id` gives more flexibility in the orchestrator implementation, even if most of the times `node_id == hostname`.

Reply:
In k8s, the node is the linux host (1 node == 1 kubelet).  The pods are individual Ceph services (each mon, osd, mds is a separate pod).

If the node_id/hostname are an either/or thing then I'm not too concerned, it's just a naming thing.  I'm more concerned if there are specific cases where both fields would exist and be different

Reply:
With DeepSea, we enforce some consistency in minion names to prevent some of those issues.  However, hostname sometimes means short name and sometimes means FQDN depending on who you ask.  Using `self.node` would help that the name only needs to be unique.  (I won't suggest `nodename` since there's a  DECnet reference.)  I could see where DeepSea would use minion name.

Reply:
OK -- so it should work if I just call this field `InventoryNode.name`.

Kubernetes also calls it "name" (which happens to be a FQDN).